src/liballoc/raw_vec.rs

```diff
@@ -235,7 +235,66 @@ impl<T, A: Alloc> RawVec<T, A> {
         }
     }
 
-    /// Doubles the size of the type's backing allocation. This is common enough
+    /// Grows the vector capacity by `capacity_increase`.
+    ///
+    /// It allows implementing amortized O(1) `push` on vector-like containers.
+    ///
+    /// # Attributes
+    ///
+    /// - `#[inline]`: LLVM is able to completely elide memory allocations in
+    ///   many cases if it can "see" where memory is allocated and freed.
+    /// - `#[cold]`: calling this function is a "rare" event
+    ///
+    #[inline]
+    #[cold]
+    pub fn grow_by(&mut self, capacity_increase: usize) {
+        let elem_size = mem::size_of::<T>();
+        assert!(elem_size != 0, "RawVecs of zero-sized types can't grow");
+
+        let (new_cap, uniq) = match self.current_layout() {
+            Some(cur) => {
+                // The invariant `elem_size * self.cap <= isize::MAX` is
+                // maintained by `alloc_guard`; the alignment will never be too
+                // large as to "not be specifiable" (so we can use
+                // `from_size_align_unchecked`).
+                let new_cap = Self::suitable_capacity(self.cap, capacity_increase);
+                let new_size = new_cap * elem_size;
+                let new_layout = unsafe {
+                    Layout::from_size_align_unchecked(new_size, cur.align())
+                };
+                let (_, usable_size) = self.a.usable_size(&new_layout);
+                let new_layout = unsafe {
+                    Layout::from_size_align_unchecked(usable_size, cur.align())
+                };
+                alloc_guard(usable_size);
+                let ptr_res = unsafe { self.a.realloc(self.ptr.as_ptr() as *mut u8,
+                                             cur,
+                                                      new_layout) };
+                match ptr_res {
+                    Ok(ptr) => (new_cap, unsafe {
+                        Unique::new_unchecked(ptr as *mut T)
+                    }),
+                    Err(e) => self.a.oom(e),
+                }
+            }
+            None => {
+                let new_cap = Self::suitable_capacity(self.cap, capacity_increase);
+                let align = mem::align_of::<T>();
+                let new_size = new_cap * elem_size;
+                let new_layout = unsafe { Layout::from_size_align_unchecked(new_size, align) };
+                let (_, new_cap) = self.a.usable_size(&new_layout);;
+                alloc_guard(new_cap);
+                match self.a.alloc_array::<T>(new_cap) {
+                    Ok(ptr) => (new_cap, ptr),
+                    Err(e) => self.a.oom(e),
+                }
+            }
+        };
+        self.ptr = uniq;
+        self.cap = new_cap;
```

Comment:
(Whole function) Tightly cuddled `unsafe { }` doesn't make the code just outside of that "safe": it all needs to add up — every calculation and comparison in this function is critical for memory safety.

I'd prefer to keep the original style of a large unsafe block around it all. In particular, inline `unsafe { ... }` expressions are not really idiomatic.

This function needs to check for overflow in the capacity number. It's a harder job now that it no longer doubles and we can't rely on the isize::MAX trick documented in the double function.

Reply:
If the capacity number overflows in `suitable_capacity` then it will panic. If the `elem_size * capacity` operation overflows then that will panic. Or what am I missing?

Reply:
I mean, the new capacity can be bigger than `isize::MAX` on 32-bit archs, and the usable size of the allocator might make it even bigger, but later there is an `alloc_guard` check that will make it panic anyways.

Reply:
I don't see where the check or explicit panic is for that? I mean, plain a + b panics on overflow only when debug assertions are enabled. We need overflow checks that are active all the time.

Reply:
I thought unsigned integer wraparound always panic'ed... does it only panic with `debug_assert` enabled?

Reply:
Yes, they are not there in release mode. Old version of this uses quite elaborate arguments about isize::MAX and maximum allocation size, and usize::MAX to avoid most of those checks.

Maybe the API can be changed so that it doesn't need to take into account exactly arbitrary growth numbers? Otherwise we're just reimplementing reserve.

Look at reserve for what it does for overflow checks.

Reply:
Why is grow_by a new function, and how is it different from reserve? :smile:

Reply:
@bluss indeed, I think I should leave `double` and `double_in_place` as they were, add this optimizations to `reserve` and then make `Vec` use reserve instead of `double` on insertion/push. This is basically the reason why I wanted a mentor.

Reply:
The overflow arguments apply to `suitable_capacity` as well. if `cap < isize::MAX`, then `cap * 1.5` and `cap * 2` cannot overflow `usize`.

Reply:
As long as the additional capacity is a parameter to a safe function, that's an arbitrary value and it needs to be checked, not trusted.